### PR TITLE
selftest3 - heater test

### PR DESCRIPTION
**Pull request by XPila**

### CODE REVIEW

File: src/common/selftest/selftest_MINI.cpp [hunk 1]
```diff
@@ -234,10 +247,30 @@ void CSelftest::log_open() {
         serial_otp = 0;
     }
     char serial[32] = "unknown";
-    char fname[32] = "test_unknown.txt";
+    const char *suffix = "";
+    switch (m_Mask) {
+    case stmFans:
+        suffix = "_fans";
```
Comment (DRracer) [inline src/common/selftest/selftest_MINI.cpp:253]: I recommend using `static const char fans[] = "_fans"; suffix = fans;`
This makes sure the compiler won't try to load the "_fans" onto stack first, but uses the address into CPU FLASH

File: src/guiapi/include/MItem_tools.hpp [hunk 2]
```diff
@@ -90,7 +90,15 @@ class MI_TEST_HEAT : public WI_LABEL_t {
 
 public:
     MI_TEST_HEAT();
+protected:
+    virtual void click(IWindowMenu &window_menu) override;
+};
 
+class MI_TEST_FANS_fine : public WI_LABEL_t {
+    static constexpr const char *const label = N_("Test FANs fine");
```
Comment (DRracer) [inline src/guiapi/include/MItem_tools.hpp:98]: the EN wording of this message is really strange, what is the intention? Does it mean, the test of the fans went ok?

**Review by DRracer — CHANGES_REQUESTED** [review]
Please address my comments in the code

File: src/common/selftest/selftest_MINI.cpp [hunk 3]
```diff
@@ -197,26 +203,33 @@ bool CSelftest::phaseAxis(const selftest_axis_config_t *pconfig_axis, CSelftestP
...
     m_pHeater_Bed = m_pHeater_Bed ? m_pHeater_Bed : new CSelftestPart_Heater(&Config_HeaterBed);
     m_pHeater_Nozzle->Loop();
     m_pHeater_Bed->Loop();
-    if (m_pHeater_Nozzle->IsInProgress() || m_pFan1->IsInProgress()) {
-        //int p0 = m_pHeater_Nozzle->GetProgress();
-        //int p1 = m_pHeater_Bed->GetProgress();
+    if (m_pHeater_Nozzle->IsInProgress() || m_pHeater_Bed->IsInProgress()) {
+        int p0 = m_pHeater_Nozzle->GetProgress();
+        int p1 = m_pHeater_Bed->GetProgress();
+        int p = _PF * ((p1 > p0) ? p0 : p1);
```
Comment (RadekVana) [inline src/common/selftest/selftest_MINI.cpp:214]: std::min

File: src/common/selftest/selftest_MINI.cpp (hunk 1 above)
Comment (XPila) [reply src/common/selftest/selftest_MINI.cpp:253]: i do not see any benefits, all the strings constants are placed in the flash, nothing loaded onto stack.

File: src/guiapi/include/MItem_tools.hpp (hunk 2 above)
Comment (XPila) [reply src/guiapi/include/MItem_tools.hpp:99]: this is fan test with 24 measured points (normal test has 5 points). This is only for Zoltan and we will remove this menuitem in RC.

**Review by DRracer — APPROVED** [review]

File: src/common/selftest/selftest_MINI.cpp [hunk 4]
```diff
@@ -234,10 +247,30 @@ void CSelftest::log_open() {
...
+        break;
+    default:
+        break;
+    }
+    char fname[64] = "test_unknown.txt";
+    sprintf(fname, "test_unknown%s.txt", suffix);
     if (serial_otp) {
         sprintf(serial, "CZPX%.15s", serial_otp);
-        sprintf(fname, "test_CZPX%.15s.txt", serial_otp);
+        sprintf(fname, "test_CZPX%.15s%s.txt", serial_otp, suffix);
```
[the remark this answers is not in the document]
Comment (XPila) [reply src/common/selftest/selftest_MINI.cpp]: fixed

File: src/common/selftest/selftest_MINI.cpp (hunk 3 above)
Comment (XPila) [reply src/common/selftest/selftest_MINI.cpp]: fixed

**Review by RadekVana — APPROVED** [review]

**Pull request merged**